src/main/scala/cromwell/binding/Scope.scala

```diff
@@ -1,14 +1,147 @@
 package cromwell.binding
 
+import cromwell.binding.AstTools.AstNodeName
+import cromwell.parser.WdlParser.{Ast, AstList, AstNode}
+
+import scala.collection.JavaConverters._
+
+
+object Scope {
+
+  /**
+   * Utility class to generate a ScopeTree from an Ast.
+   */
+  private class ScopeTreeGenerator {
+
+    /* Keep track of the # of created Scatter nodes. */
+    var scatterCount: Int = 0
+
+    /**
+     * Retrieve the list of AstNodes from the AST body.
+     * Return empty seq if body is null or is not a list.
+     */
+    private def getChildrenList(ast: Ast): Seq[AstNode] = Option(ast.getAttribute("body")) collect { case list: AstList => list.asScala.toVector } getOrElse Seq.empty
+
+    /**
+     * Generate recursively all descendant Scopes for the workflow.
+     * Should only be called during apply with a root Ast
+     * @return Some(instance of Scope) if the node has a Scope equivalent, None otherwise
+     */
+    def generateScopeTree(node: AstNode, namespaces: Seq[WdlNamespace], tasks: Seq[Task], wdlSyntaxErrorFormatter: WdlSyntaxErrorFormatter): Option[Scope] = {
+      node match {
+        case w: Ast if w.getName == AstNodeName.Workflow =>
+          val workflowChildren = getChildrenList(w) flatMap { generateScopeTree(_, namespaces, tasks, wdlSyntaxErrorFormatter) }
+          val workflow: Workflow = Workflow(w, wdlSyntaxErrorFormatter)
+          workflow.setChildren(workflowChildren)
+          Some(workflow)
+
+        case c: Ast if c.getName == AstNodeName.Call => Some(Call(c, namespaces, tasks, wdlSyntaxErrorFormatter))
+
+        case s: Ast if s.getName == AstNodeName.Scatter =>
+          // Need to create the Scatter before generating its children to have the correct index layout order
+          val scatter = Scatter(s, scatterCount)
+          scatterCount += 1
+          val scatterChildren = getChildrenList(s) flatMap { generateScopeTree(_, namespaces, tasks, wdlSyntaxErrorFormatter) }
+          scatter.setChildren(scatterChildren)
+          Some(scatter)
+
+        /* Cases need to be added here to support new scopes */
+        case _ => None
+      }
+    }
+  }
+
+  /**
+   * Generate a Scope Tree from an Ast Node.
+   * Node must match a Scope (currently: Workflow, Scatter or Scope)
+   * @throws UnsupportedOperationException if node does not match any supported Scopes
+   * @return generated Scope Tree.
+   */
+  def generateTree(node: AstNode, namespaces: Seq[WdlNamespace], tasks: Seq[Task], wdlSyntaxErrorFormatter: WdlSyntaxErrorFormatter): Scope = {
```

Comment:
The one place this outer function gets used is immediately coercing to a Workflow, should this just return a Workflow as `generateWorkflow` or something like that? Or is it expected that this gets called in other contexts?